### PR TITLE
ARROW-11299: [Python] Fix invalid-offsetof warnings

### PR DESCRIPTION
**Pull request by cyb70289**
Use unique_ptr to hold FunctionOptions derived classes instances to fix
`invalid-offsetof` python build warnings.
    
~/arrow/python/build/temp.linux-x86_64-3.6/_compute.cpp:26034:146:
warning: offsetof within non-standard-layout type ‘__pyx_obj_7pyarrow_8_compute__CastOptions’ is undefined [-Winvalid-offsetof]
 x_type_7pyarrow_8_compute__CastOptions.tp_weaklistoffset = offsetof(struct __pyx_obj_7pyarrow_8_compute__CastOptions, __pyx_base.__pyx_base.__weakref__);

### CODE REVIEW

**Comment by cyb70289** [comment]
All debug build failed due to this change. FunctionOptions is supposed to be polymorphic, but looks it not necessary.
Would like to hear comments before moving on.
@bkietz

**Comment by bkietz** [comment]
@cyb70289 `arrow::dataset::Expression::Call::options` requires that virtual destructor since a `shared_ptr<FunctionOptions>` is used to store subclasses of FunctionOptions. Removing this virtual destructor results in a resource leak since the derived classes' (implicit) destructors are not being called, leaking data members such as `arrow::compute::SetLookupOptions::value_set` which are not destroyed.

**Comment by bkietz** [comment]
I think the best way to resolve this issue is to avoid direct instances of derived classes of FunctionOptions in Cython, since Cython inappropriately applies `offsetof`. For example, note that the warning is emitted for [MinMaxOptions](https://github.com/apache/arrow/blob/fdc63acdb68adade0d68e232dfcc892dc0ecf700/python/pyarrow/_compute.pyx#L670) but not for [SetLookupOptions](https://github.com/apache/arrow/blob/fdc63acdb68adade0d68e232dfcc892dc0ecf700/python/pyarrow/_compute.pyx#L732) (where the options are held by unique_ptr rather than being a direct member).

I'd recommend the following approach: rewrite `_compute.pyx::FunctionOptions` to hold a `shared_ptr[CFunctionOptions]`, and let the `cdef` classes acquire a non owning pointer to the appropriate FunctionOptions subclass. For example:

`_compute.pxd`
```python
cdef class FunctionOptions(_Weakrefable):
    cdef:
        shared_ptr[CFunctionOptions] options

    cdef const CFunctionOptions* get_options(self) except NULL
```

`_compute.pyx`
```python
cdef class _CastOptions(FunctionOptions):
    cdef:
        const CCastOptions* cast_options

    __slots__ = ()  # avoid mistakingly creating attributes

    cdef const CFunctionOptions* get_options(self) except NULL:
        return self.cast_options
```

**Comment by cyb70289** [comment]
> @cyb70289 `arrow::dataset::Expression::Call::options` requires that virtual destructor since a `shared_ptr<FunctionOptions>` is used to store subclasses of FunctionOptions. Removing this virtual destructor results in a resource leak since the derived classes' (implicit) destructors are not being called, leaking data members such as `arrow::compute::SetLookupOptions::value_set` which are not destroyed.

Thanks @bkietz , I missed this point.
Will update according to your comment.

File: python/pyarrow/_compute.pyx [hunk 1]
```diff
@@ -481,88 +481,89 @@ cdef class FunctionOptions(_Weakrefable):
...
     def _set_type(self, target_type=None):
         if target_type is not None:
-            self.options.to_type = (
+            deref(self.options).to_type = (
                 (<DataType> ensure_type(target_type)).sp_type
             )
 
     def _set_safe(self):
-        self.options = CCastOptions.Safe()
+        self.options.reset(new CCastOptions(CCastOptions.Safe()))
```
Comment (cyb70289) [inline python/pyarrow/_compute.pyx:514]: Any way to assign `self.options` directly?

File: python/pyarrow/_compute.pyx [hunk 2]
```diff
@@ -481,88 +481,89 @@ cdef class FunctionOptions(_Weakrefable):
...
+        return self.options.get()
 
     def _set_options(self, DataType target_type, allow_int_overflow,
                      allow_time_truncate, allow_time_overflow,
                      allow_float_truncate, allow_invalid_utf8):
+        self.options.reset(new CCastOptions())
         self._set_type(target_type)
         if allow_int_overflow is not None:
-            self.allow_int_overflow = allow_int_overflow
+            deref(self.options).allow_int_overflow = allow_int_overflow
```
Comment (cyb70289) [inline python/pyarrow/_compute.pyx:497]: Looks `options` is missed in current code.

Comment (bkietz) [reply python/pyarrow/_compute.pyx:497]: I'm not sure what you mean here. Note that `self.allow_int_overflow = ` is a property setter which modifies the corresponding field in CCastOptions, so substituting `deref(self.options).allow_int_overflow = ` here is unnecessary

File: python/pyarrow/_compute.pyx (hunk 1 above)
Comment (bkietz) [reply python/pyarrow/_compute.pyx:514]: I think this is the most efficient way to write this in cython.

File: python/pyarrow/_compute.pyx (hunk 2 above)
Comment (cyb70289) [reply python/pyarrow/_compute.pyx]: Restored to original code.

**Review by bkietz — APPROVED** [review]
LGTM, thanks!

**Pull request closed without merging**